i18n-helpers/src/bin/mdbook-xgettext.rs

```diff
@@ -229,4 +425,573 @@ mod tests {
+        let (ctx, _tmp) = create_render_context(&[
+            ("book.toml", "[book]"),
+            (
+                "src/SUMMARY.md",
+                "# Summary\n\n\
+                 - [Intro](index.md)\n\
+                 # Foo\n\n\
+                 - [The Foo Chapter](foo.md)\n\
+                 \t- [The Bar Section](foo/bar.md)\n\
+                 \t\t- [The Baz Subsection](foo/bar/baz.md)\n\
+                 - [Foo Exercises](exercises/foo.md)",
+            ),
+            (
+                "src/index.md",
+                "# Intro to X\n\
+                 \n\
+                 In this course, you will learn X.\n",
+            ),
+            (
+                "src/foo.md",
+                "# How to Foo\n\
+                 \n\
+                 The first paragraph about Foo.\n",
+            ),
+            (
+                "src/foo/bar.md",
+                "# How to Bar\n\
+                 \n\
+                 The first paragraph about Bar.\n",
+            ),
+            (
+                "src/foo/bar/baz.md",
+                "# How to Baz\n\
+                 \n\
+                 The first paragraph about Baz.\n",
+            ),
+            (
+                "src/exercises/foo.md",
+                "# Exercises on Foo\n\
+                 \n\
+                 Try out this Foo example.\n",
+            ),
+        ])?;
+
+        let catalogs = create_catalogs(&ctx, 3)?;
+
+        let message_tuples = catalogs
+            .into_iter()
+            .map(|(file_rel_path, catalog)| (file_rel_path, msgids_from_catalog(catalog)))
+            .collect::<Vec<(PathBuf, Vec<String>)>>();
+
+        let expected_message_tuples = HashMap::from([
+            (
+                PathBuf::from("summary/summary"),
+                vec![
+                    "Intro",
+                    "Foo",
+                    "The Foo Chapter",
+                    "The Bar Section",
+                    "The Baz Subsection",
+                    "Foo Exercises",
+                ],
+            ),
+            (
+                PathBuf::from("summary/intro"),
+                vec!["Intro to X", "In this course, you will learn X."],
+            ),
+            (
+                PathBuf::from("foo/the-foo-chapter"),
+                vec!["How to Foo", "The first paragraph about Foo."],
+            ),
+            (
+                PathBuf::from("foo/the-foo-chapter/the-bar-section"),
+                vec![
+                    "How to Bar",
+                    "The first paragraph about Bar.",
+                    "How to Baz",
+                    "The first paragraph about Baz.",
+                ],
+            ),
+            (
+                PathBuf::from("foo/foo-exercises"),
+                vec!["Exercises on Foo", "Try out this Foo example."],
+            ),
+        ]);
+
+        // The message ids in a catalog do not have a constant order,
+        // so check for the presence of each one in the vector.
```

Comment:
This looks like a crucial detail! I would have expected the order to be deterministic (following the order in which you traverse sub-chapters)?

Reply:
Trying to remember my thinking from 4 month younger me~

I'm not sure what I meant, as this doesn't appear to be the case currently. We do need to iterate through the catalogs in the Map and match them with the file that they represent, though